server.go

```diff
@@ -97,16 +90,44 @@ func (s *Server) ServeHTTPS(addr, keyfile, certfile string, ready *chan bool) er
 	if s.listener, err = listenTLS(addr, keyfile, certfile); err != nil {
 		return err
 	}
+	s.tls = true
+	fmt.Printf("Listen http on %s\n", addr)
+	return s.doServe(ready)
+}
+
+func (s *Server) doServe(ready *chan bool) error {
+	// A dirty trick to associate a connection with the http.Request it
+	// contains. In "net/http/server.go", handler will be called
+	// immediately after ConnState changed to StateActive, so it's safe to
+	// loop through all elements in a channel to find a match remote addr.
+	q := make(chan net.Conn, 10)
 
-	// Set up server
 	proxy := http.HandlerFunc(
 		func(w http.ResponseWriter, req *http.Request) {
+			for c := range q {
```

Comment:
It should be very rare that the queue contains other connections, a length of 2 or 3 should be enough. It's similar to the spin lock OS using I think. Maybe leave it here until we saw performance issue in benchmark?